metricbeat/helper/prometheus/prometheus.go

```diff
@@ -76,110 +86,224 @@ func (p *prometheus) GetFamilies() ([]*dto.MetricFamily, error) {
 	}
 
 	families := []*dto.MetricFamily{}
-	for {
-		mf := &dto.MetricFamily{}
-		err = decoder.Decode(mf)
-		if err != nil {
-			if err == io.EOF {
-				break
+	var wg sync.WaitGroup
+	var decoderMutex = &sync.Mutex{}
+	var familiesMutex = &sync.Mutex{}
+	var done = false
+	var count = 0
+	for !done && count < maxGetFamiliesSubFuns {
+		wg.Add(1)
+		count++
+		go func(id int) {
+			for !done {
+				mf := &dto.MetricFamily{}
+				decoderMutex.Lock()
+				err = decoder.Decode(mf)
+				decoderMutex.Unlock()
+				if err != nil {
+					if err == io.EOF {
+						done = true
+					}
+				} else {
+					if familyPrefix == nil {
+						familiesMutex.Lock()
+						families = append(families, mf)
+						familiesMutex.Unlock()
```

Comment:
same thing. can return here, and then you dont need 'else' in line 112.

Reply:
No, if I would return here then I would process just one family. Each thread keeps running until all families are decoded.
If prefix is not defined always append, otherwise only append families that match the prefix.